README.md
### Elven Mint v4 (Work in progress)

NFT minting on the Elrond blockchain using [erdjs](https://github.com/ElrondNetwork/elrond-sdk-erdjs) SDK. It requires deployed smart contract which implements `esdt_nft_create`.

**Be aware that it automatically mints in a loop** Its purpose is to test stuff or for private use, not for NFT launches for sale. If time permits, there will be more valuable tools and adapted smart contracts for that.

Smart Contract example which works well with current version of the tool: [elven-nft-minter-sc](https://github.com/juliancwirko/elven-nft-minter-sc)

[Quick walkthrough video](https://youtu.be/ppvHPzwr-X4)

This is a documentation for v4

- for v3 check [here](https://github.com/juliancwirko/elven-mint/releases/tag/v3.0.0)
- for v2 check [here](https://github.com/juliancwirko/elven-mint/releases/tag/v2.0.0).
- for v1 check [here](https://github.com/juliancwirko/elven-mint/releases/tag/v1.1.0).

**Please always check and try to understand the code and possible limitations and bugs before using it for actual products**

The script relays on [nft-art-maker](https://github.com/juliancwirko/nft-art-maker) output metadata.json file. Check out the repo, and you'll find more info about it.

It would also be possible to use it without the nft-art-maker, but you need to prepare a similar metadata.json file as [here](https://github.com/juliancwirko/nft-art-maker#output-type-configuration); you would also need to pack the files into ipfs car file and upload them. Files in the car archive example: `1.png, 1.json, 2.png, 2.json ....`.

#### How it works

First of all elven-mint script assumes that you have deployed a smart contract which implements the `esdt_nft_create`. (See the link to example!) The createNft function should look like:

- https://github.com/juliancwirko/elven-nft-minter-sc/blob/main/src/lib.rs#L63

It also requires that you have your ESDT token (collection token/id), which it will use for NFT minting. It should also have proper roles assigned, and you'll be able to add these in the config file. Read more about the configuring ESDT for NFT minting [here](https://docs.elrond.com/developers/nft-tokens/).

You would also need to have a couple of files to be able to start:

1. Wallet Keystore JSON file. The one you'll get when creating a [web wallet](https://devnet-wallet.elrond.com) - it would be best to create a new one for the Smart Contract. But you can always recover and download the JSON file using [https://devnet-wallet.elrond.com/recover](https://devnet-wallet.elrond.com/recover) and your seed phrases. (There is no support for the PEM files here, but I plan to add it).
2. You would also need to have the `metadata.json` generated using mentioned above [nft-art-maker](https://github.com/juliancwirko/nft-art-maker).
3. Finally, you would need a config file `.elvenmintrc` where you need to provide a couple of crucial pieces of information. Below you'll find an example.

**Important! Keep the files private, especially password and Keystore wallet files. To avoid any mistakes, you would need to write a wallet password with every script usage.**

So your directory should look similar to this:
```
metadata.json
wallet.json
.elvenmintrc
```

In this place, you will call the `npx elven-mint` command. Or you can also install it globally `npm install -g elven-mint` and then call `elven-mint`.

#### Configuration example

.elvenmintrc file with examples:
```
{
  "smartContractAddress": "erd1qqqqqqqqqqqqqpgqqfxkus66fkjpjuxlqnhzqwk5ru9h5smwvafsy9p3g0",
  "collectionTokenId": "ELO-5f0be3",
  "initialPriceOfNFT": "100000000000000000",
  "royaltiesCut": '10',
  "currentChain": "devnet",
  "walletFileName": "wallet.json",
  "createNftFunctionName": "createNft",
  "metadataSummaryFileName": "metadata.json",
}
```

- `smartContractAddress` - Smart Contract which mints NFTs [default: undefined]
- `collectionTokenId` - Your created ESDT token for NFT collection - should also have all required roles [default: undefined]
- `initialPriceOfNFT` - Price after mint (1 egld has 18 zeros) [default: undefined]
- `royaltiesCut` - royalties cut (format is a value between '0' an '10000' (0 meaning 0% and 10000 meaning 100%) [default: '0']
- `currentChain` - Chain to be used (local, devnet, testnet, mainnet) [default: devnet]
- `walletFileName` - Your wallet json file name [default: wallet.json]
- `createNftFunctionName` - Create NFT function name in your smart contract [default: createNft]
- `metadataSummaryFileName` - Your metadata.json file name (generated by nft-art-maker) [default: metadata.json]
- `claimTokensAfterMint` - This option will allow claiming tokens after minting. They will be sent to the caller's wallet after Smart Contract mints them. When false, they will stay on the Smart Contract [default: false]

#### Handling unexpected termination of the minting process

It is crucial to be able to resume minting from the place where the process ended. It isn't possible yet. But there is a manual way of doing that. There is a helper which will console log the last sent transaction on every exit, and it looks like: 

```
elven-mint finished!
Last sent transaction:  {
  no: 2,
  name: '#2',
  txHash: '640232838d50e02102e279ec71b892e9d12aaddc85476a094001af38942fb628'
}
```

With this info, you can double-check all in the blockchain explorer, and you can modify your metadata.json file accordingly. So, for example, you can remove all first editions for which transactions are already on-chain and then restart the script. 

I plan to implement a resume script, but this isn't so crucial for now.

#### Updates

You can always install it globally if you are not sure which version you use. Do: `npm install elven-mint -g`

You can always check which version you have by: `elven-mint -v`

#### Contact me

- [https://twitter.com/JulianCwirko](https://twitter.com/JulianCwirko)


CHANGELOG.md
### [4.0.1](https://github.com/juliancwirko/elven-mint/releases/tag/v4.0.1) (2022-01-10)
- fix versions for prompt and bignumber

### [4.0.0](https://github.com/juliancwirko/elven-mint/releases/tag/v4.0.0) (2021-12-26)
- there is no more list of metadata CIDs needed - this change is because of the nft-art-maker changes. From now on, we will use only base CID for all image files and the metadata.json. Please see nft-art-maker changelog.

### [3.0.0](https://github.com/juliancwirko/elven-mint/releases/tag/v3.0.0) (2021-12-10)
- pass royalties as BigUint
- pass option to auto claim minted tokens (it requires proper smart contract)
- when there is no ipfs cid usage, the metadata will be hardcoded on-chain, it will also be encoded using base64
- adjusted to work with the example smart contract

### [2.0.0](https://github.com/juliancwirko/elven-mint/releases/tag/v2.0.0) (2021-11-19)
- changes to align it to the new nft-art-maker v3
- breaking changes in the configuration structure

### [1.1.0](https://github.com/juliancwirko/elven-mint/releases/tag/v1.1.0) (2021-11-01)
- added process state catch after exit - standard or unexpected

### [1.0.0](https://github.com/juliancwirko/elven-mint/releases/tag/v1.0.0) (2021-10-31)
- initial code


src/utils.ts
import { readFileSync, accessSync, constants } from 'fs';
import { exit, cwd } from 'process';

import {
  NetworkConfig,
  chooseProvider,
  UserSigner,
  Account,
  Address,
  IProvider,
  SmartContract,
  ContractFunction,
  GasLimit,
  BytesValue,
  BigUIntValue,
  ISigner,
  SystemConstants,
  BooleanValue,
} from '@elrondnetwork/erdjs';
import ora from 'ora';
import prompt from 'prompt';
import BigNumber from 'bignumber.js';

import { ProcessState } from './catch-on-exit';
import * as config from './config';

export interface Metadata {
  name: string;
  description: string;
  properties: {
    edition: number;
    attributes: {
      trait_type: string;
      value: string;
    }[];
    base64SvgDataUri?: string;
    hash?: string;
    tags?: string;
  };
  image: {
    href: string;
    hash: string;
    ipfsCid?: string;
    ipfsUri?: string;
    fileName?: string;
  };
}

export const baseDir = cwd();

export const getWalletPassword = async () => {
  const promptSchema = {
    properties: {
      password: {
        description: 'Wallet Keystore password',
        required: true,
        hidden: true,
        replace: '*',
      },
    },
  };

  prompt.start();

  try {
    const { password } = await prompt.get([promptSchema]);

    if (!password) {
      console.log('You have to provide wallet password!');
      exit();
    }

    return password;
  } catch (e: any) {
    console.log(e.message);
    exit();
  }
};

export const getSmartContract = () => {
  if (!config.smartContractAddress) {
    console.log(
      'Please provide your Smart Contract address in the configuration file!'
    );
    exit();
  } else {
    const contract = new SmartContract({
      address: new Address(config.smartContractAddress),
    });
    return contract;
  }
};

const fileTypeMap: Record<string, string> = {
  wallet: config.walletFileName,
  metadataSummary: config.metadataFileName,
};

export const getFileContents = (
  type: 'wallet' | 'metadataSummary',
  preventExit = false
) => {
  const filePath = `${baseDir}/${fileTypeMap[type]}`;

  try {
    accessSync(filePath, constants.R_OK | constants.W_OK);
  } catch (err) {
    if (preventExit) return null;
    console.error(`No access to the ${type} JSON file!`);
    exit();
  }

  const rawFile = readFileSync(filePath);
  return JSON.parse(rawFile.toString('utf8'));
};

export const getProvider = () => {
  return chooseProvider(config.providerIds[config.currentChain]);
};

// Sync proper chain, for example, the devnet
export const syncProviderConfig = async (provider: IProvider) => {
  return NetworkConfig.getDefault().sync(provider);
};

// Prepare user signer - we need to be able to sign transactions
// using JSON wallet file and password
// It can also be configured using a pem file, but for now, this will be enough
export const prepareUserSigner = (wallet: any, walletPassword: string) => {
  const signer = UserSigner.fromWallet(wallet, walletPassword);

  return signer;
};

// Prepare main user account from the address
export const prepareUserAccount = async (address: string) => {
  return new Account(new Address(address));
};

export const getProperMetadataCid = (
  metadataCidsList: { name: string; cid: string }[],
  editionNumber: number
) => {
  return metadataCidsList.find((item) =>
    item.name.includes(editionNumber.toString())
  )?.cid;
};

export const makeTransactions = async (
  userAccount: Account,
  provider: IProvider,
  signer: ISigner,
  smartContract: SmartContract,
  createNftFunctionName: string,
  collectionTokenId: string
) => {
  const metadata: { editions: Metadata[]; metadataFilesIpfsBaseCid: string } =
    getFileContents('metadataSummary');

  const metadataString = (entry: Metadata) => {
    if (entry.image.ipfsCid) {
      return `${metadata.metadataFilesIpfsBaseCid}/${entry.properties.edition}.json`;
    }
    // Fallback when you won't use ipfs and image files (ipfs CIDs)
    // metadata fields will be hardcoded on-chain and encoded using base64
    return Buffer.from(
      JSON.stringify({
        onChainMeta: true,
        description: entry.description,
        attributes: JSON.stringify(entry.properties.attributes),
        hash: entry.image.hash, // sha256 of the real image (png or svg)
      })
    ).toString('base64');
  };

  for (const [index, entry] of metadata.editions.entries()) {
    const token = BytesValue.fromUTF8(collectionTokenId);
    const name = BytesValue.fromUTF8(entry.name);
    const uri = BytesValue.fromUTF8(
      entry.image.href || entry.properties.base64SvgDataUri || ''
    );

    const attributes = BytesValue.fromUTF8(
      `tags:${entry.properties.tags};metadata:${metadataString(entry)}`
    );
    // The best will be if your SC creates a hash from attributes
    // But anyway, the script will take it from JSON file if any
    const hash = BytesValue.fromUTF8(entry.properties.hash || '');

    const transaction = smartContract.call({
      func: new ContractFunction(createNftFunctionName),
      args: [
        token,
        name,
        uri,
        attributes,
        hash,
        new BigUIntValue(new BigNumber(config.royaltiesCut)),
        new BigUIntValue(new BigNumber(config.initialPriceOfNFT)),
        ...(config.claimTokensAfterMint ? [new BooleanValue(true)] : []),
      ],
      gasLimit: new GasLimit(SystemConstants.MIN_TRANSACTION_GAS), // gas limit - initial value
    });

    // MIN_TRANSACTION_GAS = 50_000;
    // ESDT_ISSUE_GAS_LIMIT = 60_000_000;
    // ESDT_TRANSFER_GAS_LIMIT = 500_000;
    // ESDT_NFT_TRANSFER_GAS_LIMIT = 1_000_000;
    // ESDT_BASE_GAS_LIMIT = 6_000_000;

    // TODO: check if this can be calculated in a better way?
    const esdtNftTransferGasLimit =
      SystemConstants.ESDT_NFT_TRANSFER_GAS_LIMIT.valueOf();
    const minTransactionGas = SystemConstants.MIN_TRANSACTION_GAS.valueOf();
    const gasPerDataByte = NetworkConfig.getDefault().GasPerDataByte.valueOf();
    const esdtBaseGasLimit = SystemConstants.ESDT_BASE_GAS_LIMIT.valueOf();
    const gasLimit =
      esdtNftTransferGasLimit +
      esdtBaseGasLimit +
      attributes.getLength() * minTransactionGas +
      transaction.getData().length() * gasPerDataByte;

    transaction.setGasLimit(new GasLimit(gasLimit));

    transaction.setNonce(userAccount.nonce);
    userAccount.incrementNonce();
    signer.sign(transaction);

    const spinner = ora('Processing transactions...');
    spinner.start();

    try {
      const txHash = await transaction.send(provider);
      ProcessState.set({
        no: index + 1,
        name: entry.name,
        txHash: txHash.toString(),
      });

      await transaction.awaitExecuted(provider);

      const txStatus = transaction.getStatus();

      spinner.stop();

      if (txStatus && txStatus.isInvalid()) {
        console.log(`Invalid! Tx ${index + 1}: `, txHash.toString());
        exit();
      }
      if (txStatus && txStatus.isSuccessful()) {
        console.log(`Success! Tx ${index + 1}: `, txHash.toString());
      }
      if (txStatus && txStatus.isFailed()) {
        console.log(`Failed! Tx ${index + 1}: `, txHash.toString());
        exit();
      }
    } catch (e) {
      console.log(e);
    }
  }
};


src/elven-mint.ts
import { exit } from 'process';
import * as config from './config';

import {
  syncProviderConfig,
  prepareUserAccount,
  prepareUserSigner,
  getFileContents,
  getSmartContract,
  getProvider,
  makeTransactions,
  getWalletPassword,
} from './utils';

// Main function
export const mint = async () => {
  try {
    // Prompt configuration for the password
    const password = await getWalletPassword();

    // JSON wallet file
    const wallet = getFileContents('wallet');

    // Smart contract instance - SC responsible for minting
    const smartContract = getSmartContract();

    // Provider type based on initial configuration
    const provider = getProvider();
    await syncProviderConfig(provider);

    const userAccount = await prepareUserAccount(wallet.bech32);
    await userAccount.sync(provider);

    const signer = prepareUserSigner(wallet, password as string);

    if (
      config.collectionTokenId &&
      config.createNftFunctionName &&
      config.initialPriceOfNFT
    ) {
      await makeTransactions(
        userAccount,
        provider,
        signer,
        smartContract,
        config.createNftFunctionName,
        config.collectionTokenId
      );
    } else {
      console.log(
        'Please check if you configured your collectionTokenId, createNftFunctionName and initialPriceOfNFT'
      );
      exit();
    }
  } catch (e: any) {
    console.log('Oops, something went wrong: ', e.message);
  }
};


src/catch-on-exit.ts
import process from 'process';

interface ProcessStateType {
  no: number; // number of the processed transaction, index of the loop + 1
  name: string; // name of the edition defined in metadata.json file
  txHash: string; // transaction hash of the transaction on chain
}

export class ProcessState {
  static processState: ProcessStateType;
  static set(data: ProcessStateType) {
    this.processState = data;
  }
  static get() {
    return this.processState;
  }
}

export const initCatchOnExit = () => {
  function exitHandler() {
    const state = ProcessState.get();
    console.log(' ');
    console.log('=======================');
    console.log(' ');
    console.log('elven-mint finished!');
    console.log('Last sent transaction: ', state || 'none');
    console.log(' ');
    console.log('=======================');
    console.log(' ');
  }

  //do something when app is closing
  process.on('exit', exitHandler);

  //catches ctrl+c event
  process.on('SIGINT', exitHandler);

  // catches "kill pid" (for example: nodemon restart)
  process.on('SIGUSR1', exitHandler);
  process.on('SIGUSR2', exitHandler);

  //catches uncaught exceptions
  process.on('uncaughtException', exitHandler);
};


src/config.ts
import { cosmiconfigSync } from 'cosmiconfig';
import { cwd } from 'process';

const explorerSync = cosmiconfigSync('elvenmint');
const customConfig = explorerSync.search(cwd());

// Minter Smart Contract
export const smartContractAddress = customConfig?.config?.smartContractAddress;

// Your created ESDT token for NFT collection - should also have all required roles
export const collectionTokenId = customConfig?.config?.collectionTokenId;

// Price after mint - 1 egld has 18 zeros
export const initialPriceOfNFT = customConfig?.config?.initialPriceOfNFT;

// royalties cut (format is a numeric value between 0 an 10000 (0 meaning 0% and 10000 meaning 100%)
export const royaltiesCut = customConfig?.config?.royaltiesCut || '0';

// Chain to be used (local, devnet, testnet, mainnet)
export const currentChain = customConfig?.config?.currentChain || 'devnet';

// Your wallet json file name
export const walletFileName =
  customConfig?.config?.walletFileName || 'wallet.json';

// Create NFT function name in your smart contract
export const createNftFunctionName =
  customConfig?.config?.createNftFunctionName || 'createNft';

// Your metadata.json file name (generated by 'nft-art-maker generate')
export const metadataFileName =
  customConfig?.config?.metadataFileName || 'metadata.json';

// This option will allow claiming tokens after minting. They will be sent to the caller's wallet after mint
export const claimTokensAfterMint =
  customConfig?.config?.claimTokensAfterMint || false;

export const providerIds: { [key: string]: string } = {
  local: 'local-testnet',
  testnet: 'elrond-testnet',
  devnet: 'elrond-devnet',
  mainnet: 'elrond-mainnet',
};
